tests/rdma/run.py

```diff
@@ -0,0 +1,195 @@
+    rdmapath = valkeydir + "/src/valkey-rdma.so"
+    svrcmd = [svrpath, "--port", "0", "--loglevel", "verbose", "--protected-mode", "no",
+             "--appendonly", "no", "--daemonize", "no", "--dir", "tests/rdma/tmp",
+             "--loadmodule", rdmapath, "port=6379", "bind=" + ipaddr]
+
+    svr = subprocess.Popen(svrcmd, shell=False, stdout=subprocess.PIPE)
+    try:
+        if svr.wait(1):
+             print("Valkey Over RDMA valkey-server runs less than 1s [FAILED]")
+             print("---------------\n" + svrcmd + "---------------\n")
+             return 1
+    except subprocess.TimeoutExpired as e:
+        print("Valkey Over RDMA valkey-server start [OK]")
+        pass
+
+    # step 2, run test client
+    start = time.time()
+    clipath = valkeydir + "/tests/rdma/tmp/valkey-rdma-test"
+    clicmd = [clipath, "--thread", "8", "-h", ipaddr]
+    cli = subprocess.Popen(clicmd, shell=False, stdout=subprocess.PIPE)
+    if cli.wait(60):
+        outs, _ = cli.communicate()
+        print("Valkey Over RDMA test [FAILED]")
+        print("---------------\n" + outs.decode() + "---------------\n")
+        retval = 1
+    else:
+        elapsed = time.time() - start
+        outs, _ = cli.communicate()
+        print("Valkey Over RDMA test in " + str(round(elapsed, 2)) + "s [OK]")
+        print(outs.decode())
+        retval = 0
+
+    # step 3, cleanup
+    svr.kill()
+    svr.wait()
+    subprocess.Popen("rm -f tests/rdma/tmp/dump.rdb", shell=True).wait()
+
+    # step 4, report result
+    return retval
+
+
+if __name__ == "__main__":
+    # prepare test directory
+    subprocess.Popen("mkdir -p tests/rdma/tmp", shell=True).wait()
+
+    # build C client into binary
+    build_program()
+
+    # load IB necessary kernel modules
+    prepare_ib()
+
+    # round 1: try to test over any existing RDMA device.
+    #          If test succeeds, no need to try software emulated RDMA device.
+    ipaddr = find_rdma_dev()
+    if ipaddr is None:
+        # not fatal error, continue to create software version: RXE and SIW
+        print("Valkey Over RDMA test detect existing RDMA device [FAILED]")
+    else:
+        test_rdma(ipaddr)
+        print("Valkey Over RDMA test over " + ipaddr + " [OK]")
+        os._exit(0);
+
+    # round 2: try to setup a RXE device.
+    #          Build out-out-tree RXE driver because of the lack of support from kernel.
+    prepare_rxe()
+    rxe_dev = setup_soft_rdma("rxe")
+    if rxe_dev is not None:
+        ipaddr = find_rdma_dev()
+        if ipaddr is None:
+            # a RXE device is already created, but cant't find available address
+            print("Valkey Over RDMA test detect existing RXE device [FAILED]")
+            os._exit(1);
+
+        retval = test_rdma(ipaddr)
+        subprocess.Popen("rdma link del " + rxe_dev, shell=True).wait()
+
+        if retval:
+            os._exit(1);
+
+    # remove RXE driver from kernel
+    subprocess.Popen("rmmod rdma_rxe", shell=True).wait()
```

Comment:
If we run `make rdma-test` locally, we probably don't want to load and unload kernel modules by default.

Can we separate this? Either we can do it directly in the GitHub CI yaml file, or use some separate script or an option to this script, like `--insmod-rxe`, `--rmmod-rxe`, etc. so the tests don't need to run as root.

Reply:
I tried to do these steps in the GitHub CI yaml file, and I don't know why the steps stop after `insmod` or `modprobe`. So I have to move there steps into `run.py`.

Reply:
Hi @zuiderkwast 
I push a new version, this script detects root privilege, then  to continue test or show messages. Please take a look at, thanks!

Reply:
> If we run `make rdma-test` locally, we probably don't want to load and unload kernel modules by default.
> 
> Can we separate this? Either we can do it directly in the GitHub CI yaml file, or use some separate script or an option to this script, like `--insmod-rxe`, `--rmmod-rxe`, etc. so the tests don't need to run as root.

@pizhenwei Did you try to use `rdma link` command directly without insmod/modprobe rxe? AFAIK, when ` rdma link add ` is called, the proper kernel module will be loaded automatically.

Reply:
> > If we run `make rdma-test` locally, we probably don't want to load and unload kernel modules by default.
> > Can we separate this? Either we can do it directly in the GitHub CI yaml file, or use some separate script or an option to this script, like `--insmod-rxe`, `--rmmod-rxe`, etc. so the tests don't need to run as root.
> 
> @pizhenwei Did you try to use `rdma link` command directly without insmod/modprobe rxe? AFAIK, when `rdma link add` is called, the proper kernel module will be loaded automatically.

Right, kernel handles netlink command `rdma link add rxe0 type rxe netdev IFACE`, then automatically requires user space module installing. However, `modprobe rdma_rxe`  depends on `/usr/lib/modules/KERNELVERSION/modules.dep`. Because the CI machine did not build RXE, so there is no `rdma_rxe` information in `/usr/lib/modules/KERNELVERSION/modules.dep`, so it can't be installed automatically. We have to install `rdma_rxe.ko` manually.

Reply:
> > > If we run `make rdma-test` locally, we probably don't want to load and unload kernel modules by default.
> > > Can we separate this? Either we can do it directly in the GitHub CI yaml file, or use some separate script or an option to this script, like `--insmod-rxe`, `--rmmod-rxe`, etc. so the tests don't need to run as root.
> > 
> > 
> > @pizhenwei Did you try to use `rdma link` command directly without insmod/modprobe rxe? AFAIK, when `rdma link add` is called, the proper kernel module will be loaded automatically.
> 
> Right, kernel handles netlink command `rdma link add rxe0 type rxe netdev IFACE`, then automatically requires user space module installing. However, `modprobe rdma_rxe` depends on `/usr/lib/modules/KERNELVERSION/modules.dep`. Because the CI machine did not build RXE, so there is no `rdma_rxe` information in `/usr/lib/modules/KERNELVERSION/modules.dep`, so it can't be installed automatically. We have to install `rdma_rxe.ko` from our building path manually.